### PR TITLE
OpenPGP SHA1 warnings

### PR DESCRIPTION
**Pull request by pzduniak**
Prepare the keys and tests for:
 - pgp_encrypt - sending to a SHA1 key (fake the resolving?)
 - pgp_decrypt/verify - clearsigned/detached/attached messages with SHA1 sigs signed by SHA1/SHA256 keys.

looks like a good candidate for a table test

Uploading keys was secured by Michał on the serverside.

### CODE REVIEW

File: go/engine/pgp_common.go [hunk 1]
```diff
@@ -12,21 +12,23 @@ import (
...
 }
 
 // OutputSignatureSuccessNonKeybase prints the details of successful signature verification
 // when signing key is not known to keybase.
-func OutputSignatureSuccessNonKeybase(m libkb.MetaContext, keyID uint64, signatureTime time.Time) error {
+func OutputSignatureSuccessNonKeybase(m libkb.MetaContext, keyID uint64, signatureTime time.Time, warnings libkb.HashSecurityWarnings) error {
 	arg := keybase1.OutputSignatureSuccessNonKeybaseArg{
 		KeyID:    fmt.Sprintf("%X", keyID),
 		SignedAt: keybase1.TimeFromSeconds(signatureTime.Unix()),
+		Warnings: warnings.Strings(),
```
[the remark this answers is not in the document]
Comment (pzduniak) [reply go/engine/pgp_common.go]: Apparently crypto in the GUI is Saltpack-only.

[the remark this answers is not in the document]
Comment (zapu) [reply go/engine/pgp_common.go]: was it changed? i thought I saw some PGP checkboxes in there

[the remark this answers is not in the document]
Comment (pzduniak) [reply go/engine/pgp_common.go]: Yeah, I think we commented on the same screenshot.

**Comment by zapu** [comment]
I think we should fix the `+x` chmod. I thought it just happened to be set somehow on one of the vendored files, not all of them.

File: go/client/pgp_ui.go [hunk 2]
```diff
@@ -38,6 +38,9 @@ func (p PgpUI) OutputSignatureSuccess(_ context.Context, arg keybase1.OutputSign
 		output("Signature verified. Signed by %s %s (%s).\n", un, humanize.Time(signedAt), signedAt)
 	}
 	output("PGP Fingerprint: %s.\n", arg.Fingerprint)
+	for _, warning := range arg.Warnings {
+		output("WARNING: %s\n", warning)
```
Comment (zapu) [inline go/client/pgp_ui.go]:
```suggestion
		_, _ = p.w.Write([]byte(ColorString(p.G(), "red", fmt.Sprintf("WARNING: %s\n", warning))))
```

File: go/libkb/pgp_warnings.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+func (h HashSecurityWarning) String() string {
+	switch h.kind {
+	case HashSecurityWarningSignatureHash:
+		return fmt.Sprintf("Message's contents could've been spoofed due to its signature's %s hash scheme", HashToName[h.hash])
+	case HashSecurityWarningSignersIdentityHash:
+		return fmt.Sprintf("Signer's identity could have been spoofed due to its %s hash scheme", HashToName[h.hash])
+	case HashSecurityWarningRecipientsIdentityHash:
+		return fmt.Sprintf("Recipient's identity could have been spoofed due to its %s hash scheme", HashToName[h.hash])
```
[the remark this answers is not in the document]
Comment (maxtaco) [reply go/libkb/pgp_warnings.go]: :+1: agree with @zapu

File: go/libkb/pgp_dec.go [hunk 4]
```diff
@@ -133,9 +151,25 @@ func pgpDecryptClearsign(g *GlobalContext, source io.Reader, sink io.Writer, kr
...
+				HashSecurityWarningSignatureHash,
+				digestHash,
+			),
+		)
+	}
+
 	status.IsSigned = true
 	status.Verified = true
 	status.Entity = signer
+	status.KeyID = signer.PrimaryKey.KeyId
```
Comment (zapu) [inline go/libkb/pgp_dec.go]: is this correct? messages can also be signed by signing subkeys, if available

File: go/engine/pgp_encrypt.go [hunk 5]
```diff
@@ -163,6 +179,10 @@ func (e *PGPEncrypt) Run(m libkb.MetaContext) error {
 		}
 	}
 
+	for _, warning := range e.warnings.Strings() {
+		m.Warning(warning)
```
Comment (patrickxb) [inline go/engine/pgp_encrypt.go]: Don't do this to communicate with a user, this a log message.  The fact that it shows up in the CLI is incidental.

You should store any warnings in the engine type (like `warnings` in the PGPSignEngine, but expose it by making it `Warnings`).  After calling Run on the engine, check for warnings.  Put them in a result type avdl so clients can get them.

Comment (pzduniak) [reply go/engine/pgp_encrypt.go]: OK then, I'll add a new method to the PGP UI.

**Comment by zapu** [comment]
There are still some tests failing because of the UI changes, e.g.
```
diff --git a/go/engine/pgp_encrypt_test.go b/go/engine/pgp_encrypt_test.go
index 25cb935e3d..8c98765938 100644
--- a/go/engine/pgp_encrypt_test.go
+++ b/go/engine/pgp_encrypt_test.go
@@ -153,7 +153,11 @@ func TestPGPEncryptSelfTwice(t *testing.T) {
 	trackUI := &FakeIdentifyUI{
 		Proofs: make(map[string]string),
 	}
-	uis := libkb.UIs{IdentifyUI: trackUI, SecretUI: u.NewSecretUI()}
+	uis := libkb.UIs{
+		IdentifyUI: trackUI,
+		SecretUI:   u.NewSecretUI(),
+		PgpUI:      &TestPgpUI{},
+	}
 
 	msg := "encrypt for self only once"
 	sink := libkb.NewBufferCloser()
```

It would also be good to re-run systests locally if CI can't handle them right now after this is done.

**Comment by zapu** [comment]
```
/tmp/tmp.YHBIVWx8v2 zapu@wakaba ~ » kbs pgp sign
Running `go install github.com/keybase/client/go/keybase`...Done.
▶ WARNING Running in devel mode
▶ WARNING not exportable error: Check engine error: PGPSignEngine: requires ui "PgpUI"

/tmp/tmp.YHBIVWx8v2 zapu@wakaba ~ » kbs pgp encrypt
Running `go install github.com/keybase/client/go/keybase`...Done.
▶ WARNING Running in devel mode
▶ WARNING not exportable error: Check engine error: PGPEncrypt: requires ui "PgpUI"
```

File: go/engine/pgp_warnings_test.go [hunk 6]
```diff
@@ -0,0 +1,519 @@
...
+			Sink:         clearsignOutputSink,
+			Source:       clearsignSink,
+			AssertSigned: true,
+			SignedBy:     signedBy,
+		}
+		eng := NewPGPDecrypt(alice.tc.G, arg)
+		require.NoErrorf(t, RunEngine2(m, eng), "engine failure [%s]", e.Name)
+
+		// TODO: Y2K-1334 Fix this test
+		//require.Lenf(t, eng.SignatureStatus().Warnings, e.Count, "warnings count [%s]", e.Name)
```
Comment (zapu) [inline go/engine/pgp_warnings_test.go]:
```suggestion
		require.Lenf(t, eng.SignatureStatus().Warnings, e.Count, "warnings count [%s]", e.Name)
```
we arrive at correct warning list here, just the output text has extra newline.

File: go/client/pgp_ui.go [hunk 7]
```diff
@@ -23,7 +23,12 @@ func NewPgpUIProtocol(g *libkb.GlobalContext) rpc.Protocol {
 	return keybase1.PGPUiProtocol(g.UI.GetPgpUI())
 }
 
-func (p PgpUI) OutputSignatureSuccess(_ context.Context, arg keybase1.OutputSignatureSuccessArg) error {
+func (p PgpUI) OutputPGPWarning(_ context.Context, arg keybase1.OutputPGPWarningArg) error {
+	_, err := p.w.Write([]byte(ColorString(p.G(), "red", fmt.Sprintf("WARNING: %s\n", arg.Warning))))
+	return err
```
Comment (zapu) [inline go/client/pgp_ui.go]: I don't think we care about failures to write to output stream in pgp_ui (see line 37: `_, _ = p.w.Write([]byte(s))`). But cc @patrickxb on that

**Comment by zapu** [comment]
Still fails in the commands, you need this:

```
diff --git a/go/service/pgp.go b/go/service/pgp.go
index 226e7cde39..b3f78bbecd 100644
--- a/go/service/pgp.go
+++ b/go/service/pgp.go
@@ -71,6 +71,7 @@ func (h *PGPHandler) PGPSign(ctx context.Context, arg keybase1.PGPSignArg) (err
 	earg := engine.PGPSignArg{Sink: snk, Source: src, Opts: arg.Opts}
 	uis := libkb.UIs{
 		SecretUI:  h.getSecretUI(arg.SessionID, h.G()),
+		PgpUI:     h.getPgpUI(arg.SessionID),
 		SessionID: arg.SessionID,
 	}
 	eng := engine.NewPGPSignEngine(h.G(), &earg)
@@ -108,6 +109,7 @@ func (h *PGPHandler) PGPEncrypt(ctx context.Context, arg keybase1.PGPEncryptArg)
 	uis := libkb.UIs{
 		IdentifyUI: h.NewRemoteIdentifyUI(arg.SessionID, h.G()),
 		SecretUI:   h.getSecretUI(arg.SessionID, h.G()),
+		PgpUI:      h.getPgpUI(arg.SessionID),
 		SessionID:  arg.SessionID,
 	}
 	eng := engine.NewPGPEncrypt(h.G(), earg)
```

**Pull request merged**